File: apps/island/scripts/emoji-tileset.mjs
#!/usr/bin/env node
/**
 * emoji-tileset.mjs
 *
 * Renders a set of emojis into a 16×16 sprite sheet PNG.
 *
 * Usage:
 *   node scripts/emoji-tileset.mjs [input.json] [output.png] [--cols N]
 *
 * input.json — object mapping item names to emoji characters:
 *   { "berries": "🍒", "wood": "🪵", "rocks": "🪨" }
 *   Defaults to the built-in ITEM_ICON map if not provided.
 *
 * output.png — path to write the sprite sheet (default: public/emoji-items.png)
 *
 * --cols N — number of columns in the sheet (default: 8)
 *
 * Also writes a JSON mapping alongside the PNG:
 *   { "berries": { "col": 0, "row": 0 }, "wood": { "col": 1, "row": 0 }, ... }
 * This can be added to config/tileset.json or used as a separate items sheet.
 */

import { createCanvas, registerFont } from "canvas";
import { writeFileSync, existsSync } from "fs";
import { join, dirname, basename, extname } from "path";
import { fileURLToPath } from "url";

const __dirname = dirname(fileURLToPath(import.meta.url));

// ── Default emoji map (mirrors public/client.js ITEM_ICON) ───────────────────
const DEFAULT_ITEMS = {
  berries:   "🍒",
  wood:      "🪵",
  rocks:     "🪨",
  branches:  "🌿",
  acorns:    "🌰",
  stone_axe: "🪓",
};

// ── Parse CLI args ────────────────────────────────────────────────────────────
const args = process.argv.slice(2);
let inputPath  = null;
let outputPath = join(__dirname, "..", "public", "emoji-items.png");
let columns    = 8;

for (let i = 0; i < args.length; i++) {
  if (args[i] === "--cols" && args[i + 1]) {
    columns = parseInt(args[++i], 10);
  } else if (!inputPath && args[i].endsWith(".json")) {
    inputPath = args[i];
  } else if (!args[i].startsWith("--") && args[i].endsWith(".png")) {
    outputPath = args[i];
  }
}

// ── Load item map ─────────────────────────────────────────────────────────────
let items = DEFAULT_ITEMS;
if (inputPath) {
  if (!existsSync(inputPath)) {
    process.stderr.write(`Error: input file not found: ${inputPath}\n`);
    process.exit(1);
  }
  const raw = JSON.parse((await import("fs")).readFileSync(inputPath, "utf-8"));
  items = raw;
}

const entries = Object.entries(items);
if (entries.length === 0) {
  process.stderr.write("Error: no items to render.\n");
  process.exit(1);
}

// ── Register Noto Color Emoji if available ────────────────────────────────────
const NOTO_PATH = "/usr/share/fonts/truetype/noto/NotoColorEmoji.ttf";
if (existsSync(NOTO_PATH)) {
  registerFont(NOTO_PATH, { family: "Noto Color Emoji" });
}

// ── Render ────────────────────────────────────────────────────────────────────
const TILE = 16;
const GAP  = 1;  // 1px gap between tiles (matches the existing tileset convention)
const STEP = TILE + GAP;

const rows   = Math.ceil(entries.length / columns);
const width  = columns * STEP - GAP;
const height = rows    * STEP - GAP;

const canvas = createCanvas(width, height);
const ctx    = canvas.getContext("2d");

// Transparent background
ctx.clearRect(0, 0, width, height);

// Emoji rendering settings
ctx.textBaseline = "middle";
ctx.textAlign    = "center";

// Try to find a font size that fits the emoji within 14px (leaving 1px padding each side)
const FONT_SIZE = 12;
ctx.font = `${FONT_SIZE}px "Noto Color Emoji", Apple Color Emoji, Segoe UI Emoji, sans-serif`;

const mapping = {};

entries.forEach(([name, emoji], index) => {
  const col = index % columns;
  const row = Math.floor(index / columns);
  const cx  = col * STEP + TILE / 2;
  const cy  = row * STEP + TILE / 2 + 1; // +1 optical centering

  ctx.fillText(emoji, cx, cy);
  mapping[name] = { col, row };
});

// ── Write PNG ─────────────────────────────────────────────────────────────────
const pngBuffer = canvas.toBuffer("image/png");
writeFileSync(outputPath, pngBuffer);
process.stdout.write(`Wrote sprite sheet: ${outputPath}  (${width}×${height}px, ${entries.length} items)\n`);

// ── Write JSON mapping ────────────────────────────────────────────────────────
const base    = basename(outputPath, extname(outputPath));
const jsonOut = join(dirname(outputPath), `${base}.json`);
writeFileSync(jsonOut, JSON.stringify(mapping, null, 2));
process.stdout.write(`Wrote mapping:      ${jsonOut}\n`);
process.stdout.write(`\nMapping:\n${JSON.stringify(mapping, null, 2)}\n`);

File: apps/island/scripts/emoji-tileset.py
#!/usr/bin/env python3
"""
emoji-tileset.py

Extracts emoji bitmaps from Noto Color Emoji (CBDT) and builds a 16×16 sprite sheet.

Usage:
  python3 scripts/emoji-tileset.py [input.json] [output.png] [--cols N]

input.json  — { "berries": "🍒", ... }   (default: config/items.json)
output.png  — sprite sheet path           (default: public/emoji-items.png)
--cols N    — columns per row             (default: 8)

Also writes a JSON mapping: { "berries": {"col": 0, "row": 0}, ... }
"""

import sys, json, io, math
from pathlib import Path

# ── dependency guard ──────────────────────────────────────────────────────────
try:
    from fontTools.ttLib import TTFont
    from PIL import Image
except ImportError as e:
    print(f"Missing dependency: {e}", file=sys.stderr)
    print("Run: pip3 install fonttools pillow --break-system-packages", file=sys.stderr)
    sys.exit(1)

ROOT      = Path(__file__).resolve().parent.parent
FONT_PATH = Path("/usr/share/fonts/truetype/noto/NotoColorEmoji.ttf")
TILE      = 16
GAP       = 1
STEP      = TILE + GAP

# ── parse args ────────────────────────────────────────────────────────────────
args    = sys.argv[1:]
in_path = None
out_path = ROOT / "public" / "emoji-items.png"
columns  = 8

i = 0
while i < len(args):
    if args[i] == "--cols" and i + 1 < len(args):
        columns = int(args[i + 1]); i += 2
    elif args[i].endswith(".json"):
        in_path = Path(args[i]); i += 1
    elif args[i].endswith(".png"):
        out_path = Path(args[i]); i += 1
    else:
        i += 1

if in_path is None:
    in_path = ROOT / "config" / "items.json"

items = json.loads(in_path.read_text())
entries = list(items.items())
if not entries:
    print("Error: no items to render.", file=sys.stderr); sys.exit(1)

# ── load font & build glyph map ───────────────────────────────────────────────
if not FONT_PATH.exists():
    print(f"Error: Noto Color Emoji not found at {FONT_PATH}", file=sys.stderr)
    sys.exit(1)

font  = TTFont(str(FONT_PATH))
cmap  = font.getBestCmap()
cbdt  = font['CBDT']
# Strike 0 is the only strike (109ppem)
strike_data = cbdt.strikeData[0]

def emoji_to_glyphs(emoji):
    """Return list of glyph names for the emoji (handles ZWJ sequences)."""
    # Strip variation selectors (U+FE0F)
    codepoints = [ord(c) for c in emoji if ord(c) != 0xFE0F]
    # Try ZWJ sequence first, then single codepoint
    # Noto uses underscore-joined names for sequences: u1F1E6_u1F1E8
    if len(codepoints) > 1:
        seq_name = "_".join(f"u{cp:X}" for cp in codepoints)
        if seq_name in strike_data:
            return [seq_name]
    # Single glyph
    return [f"u{cp:X}" for cp in codepoints if cmap.get(cp)]

def render_emoji(emoji) -> Image.Image:
    """Extract the 16×16 RGBA tile for the given emoji."""
    glyph_names = emoji_to_glyphs(emoji)
    for name in glyph_names:
        gd = strike_data.get(name)
        if gd and hasattr(gd, 'imageData'):
            src = Image.open(io.BytesIO(gd.imageData)).convert("RGBA")
            return src.resize((TILE, TILE), Image.LANCZOS)
    # Fallback: grey question-mark tile
    img = Image.new("RGBA", (TILE, TILE), (100, 100, 100, 180))
    return img

# ── render all tiles ──────────────────────────────────────────────────────────
rows   = math.ceil(len(entries) / columns)
width  = columns * STEP - GAP
height = rows    * STEP - GAP

sheet = Image.new("RGBA", (width, height), (0, 0, 0, 0))

mapping = {}
for idx, (name, emoji) in enumerate(entries):
    col = idx % columns
    row = idx // columns
    tile = render_emoji(emoji)
    x = col * STEP
    y = row * STEP
    sheet.paste(tile, (x, y), tile)
    mapping[name] = {"col": col, "row": row}
    print(f"  [{idx+1}/{len(entries)}] {name:15s} {emoji}  → ({col}, {row})")

# ── write outputs ─────────────────────────────────────────────────────────────
out_path.parent.mkdir(parents=True, exist_ok=True)
sheet.save(str(out_path), "PNG")
json_out = out_path.with_suffix(".json")
json_out.write_text(json.dumps(mapping, indent=2))

print(f"\nWrote sprite sheet: {out_path}  ({width}×{height}px, {len(entries)} items)")
print(f"Wrote mapping:      {json_out}")
